examples/shader/shader_material.rs

```diff
@@ -48,91 +40,24 @@ fn setup(
     });
 }
 
-// This is the struct that will be passed to your shader
-#[derive(Debug, Clone, TypeUuid)]
-#[uuid = "f690fdae-d598-45ab-8225-97e2a3f056e0"]
-pub struct CustomMaterial {
-    color: Color,
-}
-
-#[derive(Clone)]
-pub struct GpuCustomMaterial {
-    _buffer: Buffer,
-    bind_group: BindGroup,
-}
-
-// The implementation of [`Material`] needs this impl to work properly.
-impl RenderAsset for CustomMaterial {
-    type ExtractedAsset = CustomMaterial;
-    type PreparedAsset = GpuCustomMaterial;
-    type Param = (SRes<RenderDevice>, SRes<MaterialPipeline<Self>>);
-    fn extract_asset(&self) -> Self::ExtractedAsset {
-        self.clone()
-    }
-
-    fn prepare_asset(
-        extracted_asset: Self::ExtractedAsset,
-        (render_device, material_pipeline): &mut SystemParamItem<Self::Param>,
-    ) -> Result<Self::PreparedAsset, PrepareAssetError<Self::ExtractedAsset>> {
-        let color = Vec4::from_slice(&extracted_asset.color.as_linear_rgba_f32());
-
-        let byte_buffer = [0u8; Vec4::SIZE.get() as usize];
-        let mut buffer = encase::UniformBuffer::new(byte_buffer);
-        buffer.write(&color).unwrap();
-
-        let buffer = render_device.create_buffer_with_data(&BufferInitDescriptor {
-            contents: buffer.as_ref(),
-            label: None,
-            usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
-        });
-        let bind_group = render_device.create_bind_group(&BindGroupDescriptor {
-            entries: &[BindGroupEntry {
-                binding: 0,
-                resource: buffer.as_entire_binding(),
-            }],
-            label: None,
-            layout: &material_pipeline.material_layout,
-        });
-
-        Ok(GpuCustomMaterial {
-            _buffer: buffer,
-            bind_group,
-        })
-    }
-}
-
 impl Material for CustomMaterial {
-    // When creating a custom material, you need to define either a vertex shader, a fragment shader or both.
```

Comment:
I originally added those comments because I wasn't familiar with the renderer internals and this example didn't make it obvious where the default vertex shader is coming from. As far as I know this is still the case that not specifying a vertex_shader will give you the vertex_shader of mesh.wgsl.

Did you remove it because it's not the place for it or because it isn't true anymore?

Reply:
This removal was intentional. Most of this is still true (although no custom shaders is also an option ... and at that point why list out every permutation). But I removed it all because this is intended to be a minimal example, not a tutorial that covers everything. Imo this forces too much context to the surface and changes the first impression from "look how easy this is" to "read this giant body of text providing extra context and disclaimers".

 Imo this level of detail is better suited to rustdocs and the Bevy Book.

Reply:
That being said, I do think this example would benefit from a single line comment explaining that Material has a lot of optional / default functionality (maybe with a call to action to check out the docs).

Reply:
I agree that this might not be ideal, but I personally learned bevy by looking at the examples and the ones having descriptive comments explaining what is actually going on in the examples helped me a lot. Should this be in a separate shader example that is more of a tutorial? At least as a stop-gap measure until we have a rendering chapter in the book?

Reply:
Although, I guess now that this is massively simplified it's a lot easier to figure out what is going on. I'm also super biased in the sense that I have spent the last month diving a lot deeper into wgpu/graphics programming and learning the internals of bevy so it's hard to say if I would have understood that a few months ago.

Reply:
I'd be down for a separate longer form / more complete "shader material tutorial" example (similar to the ecs one we have). But I also think that ultimately that type of thing belongs in the Bevy Book.

Examples should be small / minimal / illustrate one thing / be free of exposition beyond the minimal amount required to understand the example.